_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,323 @@
+
+    public static void main(String[] args) {
+        new Producer().run();
+    }
+
+    public void run() {
+        System.out.println("Running producer");
+
+        // Create a Kafka producer instance
+        // This producer sends messages to the Kafka topic asynchronously
+        try (var producer = createKafkaProducer()) {
+            // Generate a random byte array as the message payload
+            byte[] value = randomBytes(MESSAGE_SIZE_BYTES);
+            while (messageCount.get() < NUM_MESSAGES) {
+                sleep(PROCESSING_DELAY_MS);                
+                // Send a message to the Kafka topic, specifying topic name, message count, and message value
+                producer.send(new ProducerRecord<>(TOPIC_NAME, messageCount.get(), value), this);
+                messageCount.incrementAndGet();
+            }
+        }
+    }
+
+    private KafkaProducer<Long, byte[]> createKafkaProducer() {
+        // Create properties for the Kafka producer
+        Properties props = new Properties();
+        
+        // Configure the connection to Kafka brokers
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVERS);
+
+        // Set a unique client ID for tracking
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "client-" + UUID.randomUUID());
+        
+        // Configure serializers for keys and values
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, LongSerializer.class);
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+
+        return new KafkaProducer<>(props);
+    }
+
+    private void sleep(long ms) {
+        try {
+            TimeUnit.MILLISECONDS.sleep(ms);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private byte[] randomBytes(int size) {
+        if (size <= 0) {
+            throw new IllegalArgumentException("Record size must be greater than zero");
+        }
+        byte[] payload = new byte[size];
+        for (int i = 0; i < payload.length; ++i) {
+            payload[i] = (byte) (RND.nextInt(26) + 65);
+        }
+        return payload;
+    }
+
+    private boolean retriable(Exception e) {
+        if (e == null) {
+            return false;
+        } else if (e instanceof IllegalArgumentException
+                || e instanceof UnsupportedOperationException
+                || !(e instanceof RetriableException)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public void onCompletion(RecordMetadata metadata, Exception e) {
+        if (e != null) {
+            // If an exception occurred while sending the record
+            System.err.println(e.getMessage());
+
+            if (!retriable(e)) {
+                // If the exception is not retriable, print the stack trace and exit
+                e.printStackTrace();
+                System.exit(1);
+            }
```

Comment:
Should we retry sending this message if the error is retriable?

Reply:
@fvaleri - thoughts? We could expand the method or just add a comment about expanding it.

Reply:
Thanks @fvaleri . I updated the description of the `retriable` method

> **`retriable` method**
> 
> - Determines whether to retry sending a message following a message sending error.
> - Returns `true` if the message sending process can be retried. The Kafka producer automatically handles retries for certain errors, such as connection errors.
> - Returns `false` for null and specified exceptions, or those that do not implement the `RetriableException` interface.
> - Customizable to include other errors and  implementing retry logic for business level exceptions.

Reply:
The simple `retriable` method here is used to know when the error is fatal and we need to exit. Otherwise we simply print the error and continue, but we do not retry. This is to keep the example simple, but we can add a note above `if (!retriable(e)) {` line saying that the user should add retry logic for business errors.